### PR TITLE
Add vtr-optimized, with PGO enabled.

### PR DESCRIPTION
**Pull request by HackerFoo**
This PR adds a version of VTR that is optimized using profile-guided optimization, which reduces runtime by ~8%.

To do this, it must download a sample architecture and circuit, and run all steps to gather profile data, and then recompile using the profile data collected.

This PR supersedes https://github.com/SymbiFlow/conda-packages/pull/143

### CODE REVIEW

File: pnr/symbiflow-vtr-optimized/run-sf.sh
Comment (litghost) [inline pnr/symbiflow-vtr-optimized/run-sf.sh:25]: I'm not sure I understand this comment?  Can you explain what this working around?

File: pnr/symbiflow-vtr-optimized/build.sh
Comment (litghost) [inline pnr/symbiflow-vtr-optimized/build.sh:56]: What needs m4?

File: pnr/symbiflow-vtr-optimized/run-sf.sh
Comment (HackerFoo) [reply pnr/symbiflow-vtr-optimized/run-sf.sh:25]: Otherwise, `symbiflow_generate_constraints` in `symbiflow_place` fails with:

```
Traceback (most recent call last):
  File ".../symbiflow/share/symbiflow/scripts/prjxray_create_ioplace.py", line 178, in <module>
    main()
  File ".../symbiflow/share/symbiflow/scripts/prjxray_create_ioplace.py", line 169, in main
    io_place.output_io_place(args.output)
  File ".../symbiflow/share/symbiflow/scripts/vpr_io_place.py", line 151, in output_io_place
    max_name_length = max(len(c.name) for c in self.constraints.values())
```

File: pnr/symbiflow-vtr-optimized/build.sh
Comment (HackerFoo) [reply pnr/symbiflow-vtr-optimized/build.sh:56]: This was copied from `symbiflow-vtr/build.sh`. I'll check if it is still necessary.

File: pnr/symbiflow-vtr-optimized/run-sf.sh
Comment (litghost) [reply pnr/symbiflow-vtr-optimized/run-sf.sh:25]: I think you failed to copy the whole exception?

Comment (HackerFoo) [reply pnr/symbiflow-vtr-optimized/run-sf.sh:25]: Yes, sorry:

    ValueError: max() arg is an empty sequence

Comment (HackerFoo) [reply pnr/symbiflow-vtr-optimized/run-sf.sh:25]: This also allows me to remove the dependency on `python-constraint` which requires the `conda-forge` channel.

File: pnr/symbiflow-vtr-optimized/build.sh
Comment (HackerFoo) [reply pnr/symbiflow-vtr-optimized/build.sh]: I've removed this.

File: pnr/symbiflow-vtr-optimized/run-sf.sh
Comment (litghost) [reply pnr/symbiflow-vtr-optimized/run-sf.sh]: > This also allows me to remove the dependency on `python-constraint` which requires the `conda-forge` channel.

I believe you can get `python-constraint` from `PyPi`?

Comment (litghost) [reply pnr/symbiflow-vtr-optimized/run-sf.sh]: > Yes, sorry:
> 
> ```
> ValueError: max() arg is an empty sequence
> ```

So couple thoughts.  First, this error probably deserves to be fixed in upstream arch-defs, so it would be good to file an issue with replication instructions.  Second, I believe the fact that you hit this error likely indicates something else is wrong, not sure what.

Comment (HackerFoo) [reply pnr/symbiflow-vtr-optimized/run-sf.sh]: https://github.com/SymbiFlow/symbiflow-arch-defs/issues/1881

I think I might have found the problem:
```
ls -l benchmarks/sdc/ibex_arty.sdc
lrwxrwxrwx 1 dusty users 94 Dec 10 06:07 benchmarks/sdc/ibex_arty.sdc -> /tmpfs/src/github/symbiflow-arch-defs-continuous-install/xc/xc7/tests/soc/ibex/pins_artya7.sdc
```

Bad link.

Comment (litghost) [reply pnr/symbiflow-vtr-optimized/run-sf.sh]: I think @acomodi handles this in the VTR / symbiflow test stuff?

Comment (HackerFoo) [reply pnr/symbiflow-vtr-optimized/run-sf.sh]: https://github.com/SymbiFlow/symbiflow-arch-defs/issues/1882

Comment (HackerFoo) [reply pnr/symbiflow-vtr-optimized/run-sf.sh]: `vpr_io_place.py` still fails. If I understand correctly, the `ibex_arty.eblif` just doesn't have any constraints:

```
$ grep LOC_PAIRS symbiflow/benchmarks/circuits/ibex_arty.eblif
.param IO_LOC_PAIRS "NONE"
.param IO_LOC_PAIRS "NONE"
.param IO_LOC_PAIRS "NONE"
.param IO_LOC_PAIRS "NONE"
.param IO_LOC_PAIRS "NONE"
.param IO_LOC_PAIRS "NONE"
```

**Review by PiotrZierhoffer — APPROVED** [review]
I restarted the failing jobs (there was a problem with github connection), but from the packaging perspective - lgtm.

I am not sure if @litghost ACKs these changes as well or is there still something to be fixed?

File: pnr/symbiflow-vtr-optimized/run-sf.sh
Comment (acomodi) [reply pnr/symbiflow-vtr-optimized/run-sf.sh]: The SDC issue needs to be solved in arch-defs (https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1564#discussion_r538254173).

Regarding the IO constraints, given that the benchmarks package was originally generated to run in the VTR flow, they are already expressed as VPR placement constraints without in a `.place` file. The ibex test still uses the PCF+SDC format for constraints, hence the .eblif does not contain the IO_LOC_PAIRS.
I believe that it is worth switching to another test (e.g. picosoc or minilitex) in this PR rather then ibex, until at least we get to transition to the XDC format for all the tests that are going to be installed in the benchmark package.

**Comment by PiotrZierhoffer** [comment]
If you let me know it's good to go I'll merge right away

File: pnr/symbiflow-vtr-optimized/run-sf.sh
Comment (HackerFoo) [reply pnr/symbiflow-vtr-optimized/run-sf.sh]: I believe picosoc is too small to get sufficient profile data. I will try minilitex.

Comment (litghost) [reply pnr/symbiflow-vtr-optimized/run-sf.sh]: minilitex should work:
```
symbiflow:                                 arch.timing.xml/minilitex_ddr_arty.eblif/common                                            		OK (took 526.96 seconds)
```
from @acomodi VTR + Symbiflow test on upstream: 
https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1564
https://source.cloud.google.com/results/invocations/670bdd9e-04eb-4f0c-a667-b13421f1070d/log

Comment (HackerFoo) [reply pnr/symbiflow-vtr-optimized/run-sf.sh]: I've switched to the `minilitex_ddr_arty` design for profiling, and removed the workaround.

**Review by litghost — APPROVED** [review]
LGTM, merge if green.  Would also be good to get some concrete % improvement numbers.

File: pnr/vtr-optimized/build.sh
Comment (litghost) [inline pnr/vtr-optimized/build.sh:42]: This is not needed anymore, yes?

**Review by litghost — COMMENTED** [review]
Looks like an ibex workaround file was left in the `build.sh`?

File: pnr/vtr-optimized/build.sh
Comment (HackerFoo) [reply pnr/vtr-optimized/build.sh]: No, it's not. I've removed it.

**Comment by HackerFoo** [comment]
> Looks like an ibex workaround file was left in the `build.sh`?

I've removed it.

**Comment by litghost** [comment]
Error https://travis-ci.com/github/litex-hub/litex-conda-eda/jobs/460723313#L1245:
```
  File "/home/travis/build/litex-hub/litex-conda-eda/workdir/conda-env/lib/python3.7/site-packages/conda/resolve.py", line 288, in verify_specs
    raise ResolvePackageNotFound(bad_deps)
.
  - python-constraint
```

The key here is that you want to do a pip install, not install a conda package.  Example: https://github.com/SymbiFlow/symbiflow-arch-defs/blob/442941d382bdbe212c1a9b153e2c623aa2fd707d/environment.yml#L29
You can also include a `requirements.txt`, example: https://github.com/SymbiFlow/symbiflow-arch-defs/blob/7183cb5e59d50955686c12e4b800a55fafd1fae5/environment.yml#L24-L26

**Comment by HackerFoo** [comment]
> Error https://travis-ci.com/github/litex-hub/litex-conda-eda/jobs/460723313#L1245:
> 
> ```
>   File "/home/travis/build/litex-hub/litex-conda-eda/workdir/conda-env/lib/python3.7/site-packages/conda/resolve.py", line 288, in verify_specs
>     raise ResolvePackageNotFound(bad_deps)
> .
>   - python-constraint
> ```
> 
> The key here is that you want to do a pip install, not install a conda package. Example: https://github.com/SymbiFlow/symbiflow-arch-defs/blob/442941d382bdbe212c1a9b153e2c623aa2fd707d/environment.yml#L29
> You can also include a `requirements.txt`, example: https://github.com/SymbiFlow/symbiflow-arch-defs/blob/7183cb5e59d50955686c12e4b800a55fafd1fae5/environment.yml#L24-L26

> ValueError: Dictionaries are not supported as values in requirements sections.  Note that pip requirements as used in conda-env environment.yml files are not supported by conda-build.

**Comment by HackerFoo** [comment]
> LGTM, merge if green. Would also be good to get some concrete % improvement numbers.

For `minilitex_ddr_arty`:

| Phase | vtr | vtr-optimized | Improvement |
|--|--|--|--|
| fasm |  32.2 | 28.0 | 13% |
| pack | 99.7 | 97.2 | 2.5% |
| place | 49.3 | 43.0 | 12% |
| route | 108 | 84.9 | 22% |
| TOTAL | 289 | 253 | 13% |

**Pull request merged**
